### PR TITLE
Updated color picker clear and created a color indication in title.

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
This PR tries to close issue https://github.com/WordPress/gutenberg/issues/2770. By updating color picker to have a clear button in link style. And have a color indication in the panel titles.
## Testing

1. Add a paragraph block, choose a color for text and background, verify the color appears at the side of the title.
2. Repeat for the button block.
3. Collapse the color panels and verify the color indication persists.
4. Use the clear button and verify it clears the color changes. 

## Screenshots
screen shot 2017-10-18 at 16 19 34
screen shot 2017-10-18 at 16 20 07

### CODE REVIEW

File: components/panel/body.js [hunk 1]
```diff
@@ -52,7 +52,14 @@ class PanelBody extends Component {
 							aria-expanded={ isOpened }
 						>
 							<Dashicon icon={ icon } />
-							{ title }
+							<span className="components-panel__body-title__text-span">
+								{ title }
+							</span>
+							{ colorValue &&
```
Comment (aduth) [inline components/panel/body.js]: Is `colorValue` a concern of a general-purpose panel component? Maybe this ought to be a specialized component which extends the base panel component behaviors?

File: components/panel/style.scss [hunk 2]
```diff
@@ -59,6 +59,20 @@
 	padding: 0;
 	font-size: inherit;
 	margin-bottom: 0;
+
+	.components-panel__body-title__text-span {
+    	position: relative;
```
Comment (aduth) [inline components/panel/style.scss]: Should be tabbed once more.

Comment (jorgefilipecosta) [reply components/panel/style.scss]: 👍 Corrected

File: components/panel/body.js (hunk 1 above)
Comment (jorgefilipecosta) [reply components/panel/body.js]: Yes, I agree color is not a concern for a generic panel. I was not seeing the best way to extend and it looked like I needed to pass a prop. I end up making the panel body component receive an optional area to render at the side of the title. I created a PanelColor component that uses the panel body, rendering the correct area for the color.

**Comment by karmatosed** [comment]
Visually 👍 for the design. I think you are waiting for a code review but great work getting it like the mockups.

File: components/panel/style.scss [hunk 3]
```diff
@@ -60,6 +60,12 @@
 	font-size: inherit;
 	margin-bottom: 0;
 }
+
+.components-panel__body-title__text-span {
+	position: relative;
```
Comment (aduth) [inline components/panel/style.scss]: Why is `position: relative;` needed here or below?

File: components/panel/body.js [hunk 4]
```diff
@@ -52,7 +52,10 @@ class PanelBody extends Component {
 							aria-expanded={ isOpened }
 						>
 							<Dashicon icon={ icon } />
-							{ title }
+							<span className="components-panel__body-title__text-span">
+								{ title }
+							</span>
+							{customTitleArea}
```
Comment (aduth) [inline components/panel/body.js]: Instead of an extra `customTitleArea` prop, could the wrapping `PanelColor` component render the `title` with both the text span and color span?

```jsx
<PanelBody
	{ ...props }
	title={ [
		<span key="title">{ title }</span>,
		<span key="color" style={ { background: colorValue } }>,
	] }
/>
```

File: components/panel/body.js [hunk 5]
```diff
@@ -52,7 +52,10 @@ class PanelBody extends Component {
 							aria-expanded={ isOpened }
 						>
 							<Dashicon icon={ icon } />
-							{ title }
+							<span className="components-panel__body-title__text-span">
```
Comment (aduth) [inline components/panel/body.js]: Would recommend to avoid including an element name in a class name, since it's incidental and can change over time, isn't representative of the semantic meaning of the element.

**Comment by jorgefilipecosta** [comment]
Hi @aduth, I followed your sugestions and simplified the css, thank you :)

**Review by aduth — CHANGES_REQUESTED** [review]
Small issue: When a color value is assigned, the height of the toggle is adjusted slightly. We should try to avoid this:

resize

**Comment by mtias** [comment]
It might be nice to default to closed panels for text / background color now.

**Comment by jorgefilipecosta** [comment]
I updated the PR solving the small move identified by @aduth and I made color panels hidden by default as @mtias suggested.

File: components/panel/color.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
+/**
+ * Internal dependencies
+ */
+import './style.scss';
+import PanelBody from './body';
+
+function PanelColor( { colorValue, title, initialOpen = false, ...props } ) {
```
Comment (aduth) [inline components/panel/color.js]: @mtias Do you mean for any colored panel to default to closed, or specifically text and background color on the paragraph block?

File: components/panel/color.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+	return (
+		<PanelBody
+			{ ...props }
+			{ ...{ initialOpen } }
```
Comment (aduth) [inline components/panel/color.js]: Aside from being a bit more difficult to read, there is a potential (minor) performance impact here on iterating an additional object to extend the props of the element, vs. the more basic `initialOpen={ initialOpen }`

See:

http://babeljs.io/repl/#?babili=false&browsers=ie%2011&build=&builtIns=true&code_lz=MYewdgzgLgBADgJxHCMC8MDeMCGAuGARgBoYAjAgJhgF8BuAKFElmHSxmAIGZbGnw0GAFMANsIC2wsFADKiYTgAm7ABQMYMADwAxECA2aOAOlOJkqGoc3ZTx7Gxq1rMKAEso4tNgDaAXWdNAHoAPgYASn5mITFJaSgAORB5BEUVDHVNXX0XWzMkFECjYG9OIs13T2FS_yLQiLogA&debug=false&circleciRepo=&evaluate=false&lineWrap=false&presets=es2015%2Creact&targets=&version=6.26.0

Comment (jorgefilipecosta) [reply components/panel/color.js]: Thank you @aduth improvement applied!

File: blocks/color-palette/style.scss [hunk 8]
```diff
@@ -3,6 +3,11 @@ $color-palette-circle-spacing: 14px;
 
 .blocks-color-palette {
 	margin-right: -14px;
+
+	.blocks-color-palette__clear {
```
Comment (youknowriad) [inline blocks/color-palette/style.scss]: Should we avoid nesting here? We tend to favor top-level rules when possible.

**Review by youknowriad — APPROVED** [review]
Nice 👍

File: blocks/color-palette/style.scss (hunk 8 above)
Comment (jorgefilipecosta) [reply blocks/color-palette/style.scss]: Hi, the reason why this one is nested is for the CSS rule to have higher specificity. If the rule is not nested the margin is overwritten.

**Comment by jorgefilipecosta** [comment]
Hi @youknowriad thank you for the review!

**Comment by jorgefilipecosta** [comment]
Hi, @aduth, @mtias, @youknowriad, I added the changes to make colors collapsed in the paragraph and expanded in button. In order to make things work nicely, I added a simple change in panel body that resets the state when initialOpen prop changes. The changes are in a new commit, I tested and things are looking ok.

File: components/panel/body.js [hunk 9]
```diff
@@ -36,6 +36,14 @@ class PanelBody extends Component {
 		}
 	}
 
+	componentWillReceiveProps( nextProps ) {
+		if ( this.props.initialOpen !== nextProps.initialOpen ) {
```
Comment (youknowriad) [inline components/panel/body.js]: like its name suggest `intialOpen` should not change and if it does we should console.error. It means this panel is "uncontrolled". if we want a "controlled" panel body we can use the "opened" prop.

Can I aks why did you need this change?

Comment (jorgefilipecosta) [reply components/panel/body.js]: When changing blocks for example from a button to a paragraph. New props are passed, a new component is not built which makes sense so if you collapsed it in a block you can keep it collapsed when changing it. But if changing to different initial open settings without this the state keeps the same and the setting is useless.

Comment (youknowriad) [reply components/panel/body.js]: > a new component is not built which

I wonder why it's not the case, the block inspector of a button and a paragraph are complete separate components, I expect them to "remount" when switching between these two.

If they're not, I suppose this could create issues elsewhere. (any uncontrolled state value)

Comment (jorgefilipecosta) [reply components/panel/body.js]: I also expected them to remount but unfortunately, that is not the case, they just receive new props. Maybe there is an improvement that can make them remount but sure how.

Comment (aduth) [reply components/panel/body.js]: You can assign a unique `key` prop to trigger remounts.

Comment (youknowriad) [reply components/panel/body.js]: This is a really weird bug. I suspect some relation with slot-and-fill @aduth

Comment (aduth) [reply components/panel/body.js]: >This is a really weird bug. I suspect some relation with slot-and-fill @aduth

I don't know, it seems like React may just be recycling the same element.

Comment (aduth) [reply components/panel/body.js]: I expect if you give a unique key to `InspectorControls` here:

https://github.com/WordPress/gutenberg/blob/aee35468ebc4159a134672869cc6d69162e30d6a/blocks/library/paragraph/index.js#L114

Like `paragraph-inspector`... that it may resolve.

...but this sounds like a tricky implementation detail we shouldn't be expecting of block authors.

Comment (youknowriad) [reply components/panel/body.js]: Shouldn't we assume that `slot-fill` ensures the remounting?

Comment (youknowriad) [reply components/panel/body.js]: If it's a different fill, it should remount

Comment (jorgefilipecosta) [reply components/panel/body.js]: Hi @aduth, @youknowriad, I tried the key fix in inspector controls and I also tried adding keys to PanelColor but the components continue to be reused. It's really strange.

Comment (jorgefilipecosta) [reply components/panel/body.js]: It looks like react tries to reuse the components whatever it cans and although keys are a hint there is no guarantee a component will not be reused.

Comment (jorgefilipecosta) [reply components/panel/body.js]: It looks like the correct way is to manage state from the block components.

**Comment by jorgefilipecosta** [comment]
Hi @aduth, @youknowriad, I researched a little bit more but the keys problem that was affecting this PR. It looks like every component we rendered in fill is "copied" to the slot by the slot-fill library, during this process the keys that the elements have are ignored and new sequential ones are assigned. That's is the reason why even setting keys on ColorPanels did not manage to make the components remount.
The children of the components passed to fill keep their keys in slot, so In the InspectorControls component, I added a div around the children. So this div will have an assigned key but children (the components we pass to InspectorControls) keep their id's. I assigned unique id's to ColorPanels in paragraph and button and now things are working as they should :)

**Comment by ahmadawais** [comment]
Folks, I think the white color that appears from the center on focus is actually quite confusing and we should do away with it.

File: blocks/color-palette/index.js [hunk 10]
```diff
@@ -61,15 +61,12 @@ function ColorPalette( { colors, value, onChange } ) {
...
-			<div className="blocks-color-palette__item-wrapper blocks-color-palette__clear-color">
-				<button
-					className="blocks-color-palette__item"
-					onClick={ () => onChange( undefined ) }
-					aria-label={ __( 'Remove color' ) }
-				>
-					<span className="blocks-color-palette__clear-color-line" />
-				</button>
-			</div>
+			<button className="button-link blocks-color-palette__clear"
```
Comment (afercia) [inline blocks/color-palette/index.js]: For consistency, and in the case this component gets reused within a form element, I'd consider to add a `type="button"`

File: blocks/color-palette/index.js [hunk 11]
```diff
@@ -61,15 +61,12 @@ function ColorPalette( { colors, value, onChange } ) {
...
+				onClick={ () => onChange( undefined ) }
+				aria-label={ __( 'Remove color' ) }
```
Comment (afercia) [inline blocks/color-palette/index.js]: Since this button has now a visible text, the aria-label can be removed.

File: blocks/color-palette/index.js [hunk 12]
```diff
@@ -61,15 +61,12 @@ function ColorPalette( { colors, value, onChange } ) {
...
+			>
+				{ __( 'Clear' ) }
```
Comment (afercia) [inline blocks/color-palette/index.js]: Ideally, for better a11y, this text should be more specific and indicate _which_ color is going to be cleared. We should avoid to have multiple UI controls with the same name in the same page. Reading out of context, "Clear", "Clear", or for example "OK", "OK", "OK", wouldn't help users understand what a control does. Speech recognition software wouldn't be able to distinguish the controls. For example, when users will say the voice command "click clear", the software wouldn't be able to understand which button should be activated.
Ideally, they should be something like:
`Clear background color`
`Clear text color`
Not sure how to implement it, though.

File: blocks/library/button/index.js [hunk 13]
```diff
@@ -115,18 +115,18 @@ registerBlockType( 'core/button', {
 							checked={ !! clear }
 							onChange={ toggleClear }
 						/>
-						<PanelBody title={ __( 'Button Background Color' ) }>
+						<PanelColor title={ __( 'Button Background' ) } colorValue={ color } key="panel-color-button-1">
```
Comment (afercia) [inline blocks/library/button/index.js]: I'd recomment to don't remove the word "Color". See below. Instead, I'd consider to remove the word "Button", for consistency with the paragraphs and because it's already clear we're editing a button block.

File: blocks/library/button/index.js [hunk 14]
```diff
@@ -115,18 +115,18 @@ registerBlockType( 'core/button', {
...
 							<ColorPalette
 								value={ color }
 								onChange={ ( colorValue ) => setAttributes( { color: colorValue } ) }
 							/>
-						</PanelBody>
-						<PanelBody title={ __( 'Button Text Color' ) }>
+						</PanelColor>
+						<PanelColor title={ __( 'Button Text' ) } colorValue={ textColor } key="panel-color-button-2">
```
Comment (afercia) [inline blocks/library/button/index.js]: Removing the word "Color" doesn't help users understand what this control does. "Button Text" could be the text itself, or the font, or the text size? I'd just use `Background Color` and `Text Color` for both the paragraph and the button block.

File: blocks/library/paragraph/index.js [hunk 15]
```diff
@@ -131,18 +131,18 @@ registerBlockType( 'core/paragraph', {
 							allowReset
 						/>
 					</PanelBody>
-					<PanelBody title={ __( 'Background Color' ) }>
+					<PanelColor title={ __( 'Background' ) } colorValue={ backgroundColor } initialOpen={ false } key="panel-color-paragraph-1">
```
Comment (afercia) [inline blocks/library/paragraph/index.js]: See above, I;d recommend to not remove the word "Color".

File: blocks/library/paragraph/index.js [hunk 16]
```diff
@@ -131,18 +131,18 @@ registerBlockType( 'core/paragraph', {
...
 						<ColorPalette
 							value={ backgroundColor }
 							onChange={ ( colorValue ) => setAttributes( { backgroundColor: colorValue } ) }
 						/>
-					</PanelBody>
-					<PanelBody title={ __( 'Text Color' ) }>
+					</PanelColor>
+					<PanelColor title={ __( 'Text' ) } colorValue={ textColor } initialOpen={ false } key="panel-color-paragraph-2">
```
Comment (afercia) [inline blocks/library/paragraph/index.js]: See above, I'd recommend to not remove the word "Color".

File: components/panel/color.js [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+import PanelBody from './body';
+
+function PanelColor( { colorValue, title, ...props } ) {
+	return (
+		<PanelBody
+			{ ...props }
+			title={ [
+				<span className="components-panel__color-title" key="title">{ title }</span>,
+				colorValue && <span className="components-panel__color-area" key="color" style={ { background: colorValue } } />,
+			] }
```
Comment (afercia) [inline components/panel/color.js]: I like the overall idea to show the selected color. However, this is just a visual indication. There's no text or label that can be read by software. It would be nice to find a way to add some meaningful ext, whether it's visually hidden text or an `aria-label` attribute to override the button text.

**Comment by jorgefilipecosta** [comment]
Hi @ahmadawais, thank you for your suggestion. This PR adds a color indication in the title making it easier for the user to know which color is selected, so it mitigates the problem you are referring. Although changing the way we show selected color on the pallet is something we can discuss and try to improve in the future.

Hi @afercia thank you a lot for your inputs. I addressed most of them including making color labels be "Background Color"/"Text Color" for both paragraph and button. These ones I'm not certain if for now, we should just use this labels for aria and not display color in title label because we are already showing a color cc: @mtias. Changes are in separate commits so we can easily change them or revert parts of them if needed.
Regarding your suggestions to improve aria labels for clear buttons and provide a color indication as an aria-label, I would like to address them in a separate PR  because adding them here would make this PR big and harder to review.

**Comment by afercia** [comment]
@jorgefilipecosta thanks!
> we should just use this labels for aria and not display color in title label because we are already showing a color

Visible labels help everyone, not just assistive technologies users. I'd recommend to keep the word "Color". What was the original reasoning behind removing "Color"? Make space for the color rectangle? :) Then I'd say we're removing relevant information just for visual purposes.

**Comment by ahmadawais** [comment]
@jorgefilipecosta 

>Although changing the way we show selected color on the pallet is something we can discuss and try to improve in the future.

That's what I want.

File: blocks/color-palette/index.js (hunk 10 above)
Comment (aduth) [reply blocks/color-palette/index.js]: I also think this could be consolidated into the Button component, similar to how we have `<Button isPrimary />`, e.g. `<Button isLink />` to handle the styling and sensible defaults.

Comment (jorgefilipecosta) [reply blocks/color-palette/index.js]: Hi @aduth, unfortunately, color palette does not use button component I did a simple test to use it and in other cases, the design looks different. Also adding a setting isTypeButton to button component would break some tests. I will try to address this feedback in a separate PR, that makes color palette uses our button component.

**Comment by jorgefilipecosta** [comment]
Hi @aduth, @youknowriad given the merge of https://github.com/WordPress/gutenberg/pull/3404. I rebased and simplified this by removing all slot-fill related hacks. If you could have a last look at merging it would be perfect :) Then I will continue to iterate and address some more other a11y issues.

File: blocks/color-palette/index.js [hunk 18]
```diff
@@ -61,15 +61,11 @@ function ColorPalette( { colors, value, onChange } ) {
...
-			<div className="blocks-color-palette__item-wrapper blocks-color-palette__clear-color">
-				<button
-					className="blocks-color-palette__item"
-					onClick={ () => onChange( undefined ) }
-					aria-label={ __( 'Remove color' ) }
-				>
-					<span className="blocks-color-palette__clear-color-line" />
-				</button>
-			</div>
+			<button className="button-link blocks-color-palette__clear" type="button"
```
Comment (aduth) [inline blocks/color-palette/index.js]: I'd suggest one of two conventions for applying props to an element:

Single line

```jsx
<div className="..." onClick={ () => {} }>
	Hello World
</div>
```

Multiple lines (particularly when single line exceeds 80 characters):

```jsx
<div
	className="..."
	onClick={ () => {} }
>
	Hello World
</div>
```

Currently, this is an awkward mix of single- and multi-line style.

File: blocks/library/paragraph/index.js [hunk 19]
```diff
@@ -111,7 +111,7 @@ registerBlockType( 'core/paragraph', {
 				</BlockControls>
 			),
 			focus && (
-				<InspectorControls key="inspector">
+				<InspectorControls key="inspector" identifier="inspector-paragraph">
```
Comment (aduth) [inline blocks/library/paragraph/index.js]: What is `identifier`?

File: blocks/color-palette/index.js (hunk 18 above)
Comment (jorgefilipecosta) [reply blocks/color-palette/index.js]: Corrected 👍

File: blocks/library/paragraph/index.js (hunk 19 above)
Comment (jorgefilipecosta) [reply blocks/library/paragraph/index.js]: It was part of an experiment to solve the slot-fill keys problem. And I missed its removal it was corrected.

**Review by aduth — APPROVED** [review]
Merge conflicts, but otherwise 👍

**Comment by jorgefilipecosta** [comment]
Thank you for your review @aduth, I'm resolving the merge conflicts.
